Feature flag steps

Cawllec opened the pull request:
## Goal

Adds specific steps to be used when verify featureFlag presence in a variety of ways.

Example of usage (and practical test) can be seen here: https://buildkite.com/bugsnag/bugsnag-js-browser/builds/1745#188a45c2-791e-4477-9a5a-6677fe94bb77/493-494

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+# @step_input table [Cucumber::MultilineArgument::DataTable] Table of expected values
+Then('the event contains the following feature flags:') do |table|
+  featureFlags = Maze::Helper.read_key_path(Maze::Server.errors.current[:body], "events.0.featureFlags")
+  assert_false(featureFlags.nil? || featureFlags.empty?, "Feature flags were nil or empty: #{featureFlags}")
```
Same as above with the strictness of this

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+  assert_false(featureFlags.nil? || featureFlags.empty?, "Feature flags were nil or empty: #{featureFlags}")
+
+  expected_features = table.hashes
+  assert(featureFlags.size == expected_features.size, "Expected #{expected_features.size} features, found #{featureFlags.size}")
```
It'd be good to see what the features actually were when this fails

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+  expected_features = table.hashes
+  assert(featureFlags.size == expected_features.size, "Expected #{expected_features.size} features, found #{featureFlags.size}")
+  expected_features.each do |expected|
+    if expected['variant'].nil? || expected['variant'].empty?
+      assert(featureFlags.one? { |flag| flag['featureFlag'] == expected['featureFlag'] && flag['variant'].nil?})
```
Same thing about how strict this should be

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+Then('the event contains the feature flag {string} with variant {string}') do |flag_name, variant|
+  featureFlags = Maze::Helper.read_key_path(Maze::Server.errors.current[:body], "events.0.featureFlags")
+  assert_false(featureFlags.nil? || featureFlags.empty?, "Feature flags were nil or empty: #{featureFlags}")
+  assert(featureFlags.one? { |flag| flag['featureFlag'] == flag_name && flag['variant'] == variant })
```
Can `variant` ever be anything other than a string? I think Cucumber will always provide a string when matching `{string}`, but I'm not 100% sure

e.g. if `flag['variant']` was an integer `123`, this should fail:

```gherkin
Then the event contains the feature flag "name" with variant "123"
```

Cawllec replied:
I read in the ED that variant should always be assumed to be a string, is this not the case?

imjoehaines replied:
Yes, sorry that was what I was trying to say — it should be impossible for the `variant` variable to be anything other than a string, otherwise the `==` could pass e.g. if they were both integers

I think it's fine as-is, just wanted to check how cucumber handles `{string}` with values that could be considered non-strings (`123`/`null`/etc&hellip;)

twometresteve replied:
Cucumber is type sensitive, e.g. for :
```
{
  "values": {
    "string":"123",
    "integer":123"
  }
}

This both pass:
```
    And the error payload field "values.string" equals "123"
    And the error payload field "values.integer" equals 123
```

and these both fail:
```
    And the error payload field "values.string" equals 123
    And the error payload field "values.integer" equals "123"
```

twometresteve replied:
...so the fact that the step uses `variant {string}` and not `variant {int}` means it's ok.

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+# @step_input flag_name [String] The featureFlag value not expected
+Then('the event does not contain the feature flag {string}') do |flag_name|
+  featureFlags = Maze::Helper.read_key_path(Maze::Server.errors.current[:body], "events.0.featureFlags")
+  assert_false(featureFlags.nil? || featureFlags.empty?, "Feature flags were nil or empty: #{featureFlags}")
```
Same as above with the strictness of this

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+  assert(featureFlags.size == expected_features.size, "Expected #{expected_features.size} features, found #{featureFlags.size}")
+  expected_features.each do |expected|
+    if expected['variant'].nil? || expected['variant'].empty?
+      assert(featureFlags.one? { |flag| flag['featureFlag'] == expected['featureFlag'] && flag['variant'].nil?})
+    else
+      assert(featureFlags.one? { |flag| flag['featureFlag'] == expected['featureFlag'] && flag['variant'] == expected['variant']})
```
These assertions don't have a message when they fail; it'd be good to log the actual feature flags for debugging

imjoehaines reviewed (changes requested):
Left a bunch of comments but they're pretty much all one of two things:

1. the `event.featureFlags` has to be missing or an array; `null` isn't allowed: https://github.com/bugsnag/bugsnag-event-worker/blob/36516d2f55b72002be05c32c0b3ad252faab56ab/lib/event/event.coffee#L411-L414
2. the `variant` has to be missing or a string; it also doesn't allow `null`: https://github.com/bugsnag/bugsnag-event-worker/blob/36516d2f55b72002be05c32c0b3ad252faab56ab/lib/event/event.coffee#L427-L429

twometresteve commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,64 @@
+
+# Verifies that the are no feature flags present
+Then('the event has no feature flags') do
+  featureFlags = Maze::Helper.read_key_path(Maze::Server.errors.current[:body], "events.0.featureFlags")
```
Variable naming would be more consistent as `feature_flags` and not violate Rubocop, if we turned it on.

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+end
+
+def verify_feature_flags_with_table(event, table)
+  assert(has_feature_flags?(event), "Expected feature flags were not present in event #{event_id}: #{event}")
```
`event_id` isn't in scope here

```suggestion
  assert(has_feature_flags?(event), "Expected feature flags were not present in event: #{event}")
```

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+# Verifies that a number of feature flags outlined in a table are all present and unique in the given event
+#
+# The DataTable used for this step should have `featureFlag` and `variant` columns, containing the appropriate
+# values.  For missing or `null` variants, leave the `variant` column blank.
```
Null isn't allowed as a value, so I don't think we need to mention it here:

```suggestion
# values. For flags without a variant, leave the `variant` column blank.
```

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+#   | featureFlag | variant |
+#   | my_flag_1   | var_1   |
+#   | my_flag_2   | var_2   |
+#   | my_flag_3   |         | # Variant expected is nil
```
```suggestion
#   | my_flag_3   |         | # 'my_flag_3' should not have a variant
```

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+  event = Maze::Helper.read_key_path(Maze::Server.errors.current[:body], "events.#{event_id}")
+  assert(has_feature_flags?(event), "Expected feature flags were not present in event #{event_id}: #{event}")
+  feature_flags = event['featureFlags']
+  assert(
+    feature_flags.one? { |flag| flag['featureFlag'] == flag_name && flag['variant'] == variant },
+    "Feature flag: #{flag_name} with variant: #{variant} not found. Present flags: #{feature_flags}"
+  )
```
Is it possible to write some tests for these steps? The logic is quite complicated so it'd be good to prove they work & have a record of how they should work

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+  assert(has_feature_flags?(event), "Expected feature flags were not present in event #{event_id}: #{event}")
+  feature_flags = event['featureFlags']
+  assert(
+    feature_flags.one? { |flag| flag['featureFlag'] == flag_name && !flag.has_key?('variant') },
```
This could allow duplicates if there are multiple flags with the same `featureFlag` but only one has no `variant`

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+  expected_features.each do |expected|
+    if expected['variant'].nil? || expected['variant'].empty?
+      assert(
+        feature_flags.one? { |flag| flag['featureFlag'] == expected['featureFlag'] && !flag.has_key?('variant') },
```
same thing about duplicates

imjoehaines commented on lib/features/steps/feature_flag_steps.rb:
```diff
@@ -0,0 +1,144 @@
+      )
+    else
+      assert(
+        feature_flags.one? { |flag| flag['featureFlag'] == expected['featureFlag'] && flag['variant'] == expected['variant'] },
```
same thing about duplicates

Cawllec commented:
I believe I've addressed the duplication problem - I had to decouple some of the assertions but hopefully it still reads logically.  I've added some basic integration tests running [here](https://buildkite.com/bugsnag/maze-runner/builds/1886#f07664af-6049-42f2-953b-258419b171bd/1199) to hopefully give some confidence.
